src/sys/socket/addr.rs

```diff
@@ -1014,6 +1031,75 @@ pub mod netlink {
     }
 }
 
+#[cfg(any(target_os = "android", target_os = "linux"))]
+pub mod alg {
+    use libc::{AF_ALG, sockaddr_alg};
+    use std::{fmt, mem, str};
+    use std::hash::{Hash, Hasher};
+
+    #[derive(Copy, Clone)]
+    pub struct AlgAddr(pub sockaddr_alg);
+
+    // , PartialEq, Eq, Debug, Hash
+    impl PartialEq for AlgAddr {
+        fn eq(&self, other: &Self) -> bool {
+            let (inner, other) = (self.0, other.0);
+            (inner.salg_family, &inner.salg_type[..], inner.salg_feat, inner.salg_mask, &inner.salg_name[..]) ==
+            (other.salg_family, &other.salg_type[..], other.salg_feat, other.salg_mask, &other.salg_name[..])
+        }
+    }
+
+    impl Eq for AlgAddr {}
+
+    impl Hash for AlgAddr {
+        fn hash<H: Hasher>(&self, s: &mut H) {
+            let inner = self.0;
+            (inner.salg_family, &inner.salg_type[..], inner.salg_feat, inner.salg_mask, &inner.salg_name[..]).hash(s);
+        }
+    }
+
+    fn to_str(bytes: &[u8]) -> &str {
+        let nul_position = bytes.iter()
+            .position(|&c| c == b'\0')
+            .unwrap_or(bytes.len());
+        unsafe { str::from_utf8_unchecked(&bytes[0..nul_position]) }
+    }
+
+    impl AlgAddr {
+        pub fn new(alg_type: &str, alg_name: &str) -> AlgAddr {
+            let mut addr: sockaddr_alg = unsafe { mem::zeroed() };
+            addr.salg_family = AF_ALG as u16;
+            addr.salg_type[..alg_type.len()].copy_from_slice(alg_type.to_string().as_bytes());
+            addr.salg_name[..alg_name.len()].copy_from_slice(alg_name.to_string().as_bytes());
+
+            AlgAddr(addr)
+        }
+
+
+        pub fn alg_type(&self) -> &str {
```

Comment:
Why not simply return `&CStr` or `Result<&CStr>`?  That's what other Nix methods do.

Reply:
Changing to `CStr` should be simple. Does it make sense, considering the requirement to implement searching for string terminator by hands?

Reply:
It makes sense just for consistency's and transparency's sake, even though there is no requirement to search for the string terminator by hand.

Reply:
`CStr::from_ptr` ?